ipfs: workaround for upstream bug; other small fixes

elitak opened the pull request:
###### Motivation for this change
Temporary workaround for https://github.com/ipfs/go-ipfs/issues/4214

###### Things done
Running `ipfs repo fsck` prestart should be safe, and fixes the service not starting after an unclean exit.

Also, the ipfs autoMount now works, although I had to take the suboptimal approach of adding a line to `environment.etc."fuse.conf".text`.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

elitak committed:
ipfs: workaround for upstream bug; doc fixes

elitak committed:
ipfs: autoMount working without root

mguentner committed:
tests: ipfs: enable autoMount tests

mguentner commented:
Looks good to me now (it appears that you force pushed). Software is just great that depends on terminating newlines ;)
Could you please fetch and merge https://github.com/mguentner/nixpkgs/tree/ipfs_test_new into this PR?

elitak commented:
Oh, took me a while to get what you mean. Yeah, and I had mkIf not high up enough as well.

I've added your tests change (+ removal of "not yet" comments)  and rebased.

fpletz commented:
Thank you both! :beers:

vojta001 commented:
@elitak According to ipfs/go-ipfs#4214 (specifically [this comment](https://github.com/ipfs/go-ipfs/issues/4214#issuecomment-589231524)) the issue has been resolved. Can't we remove the `ipfs repo fsck`?

elitak commented:
@vojta001 Turns out we still need `ipfs repo fsck` because the daemon is still dying to systemd and leaving behind the `/var/lib/ipfs/api` lockfile. I can't be bothered to chase down why; trying to report bugs upstream has proven highly unproductive.

vojta001 commented:
@elitak I have no experience with ipfs developers, however, if the [release notes for go-ipfs 0.5](https://github.com/ipfs/go-ipfs/releases/tag/v0.5.0) are correct, it should truly be no longer necessary (nor available, in fact).

Proposed in #86242.
